LiteCore/Query/SQLiteDateTimeHelpers.hh

```diff
@@ -0,0 +1,232 @@
+using namespace date;
+using namespace fleece;
+
+namespace date {
+
+    using quarters    = duration<int, detail::ratio_multiply<std::ratio<3>, months::period>>;
+    using decades     = duration<int, detail::ratio_multiply<std::ratio<10>, years::period>>;
+    using centuries   = duration<int, detail::ratio_multiply<std::ratio<100>, years::period>>;
+    using millenniums = duration<int, detail::ratio_multiply<std::ratio<1000>, years::period>>;
+
+    typedef struct {
+        int64_t year;
+        int64_t doy;
+        int64_t hour;
+        int64_t minute;
+        int64_t second;
+        int64_t millisecond;
+    } DateDiff;
+
+}  // namespace date
+
+namespace litecore {
+    // A number of functions in this file are borrowed directly from the Server N1QL code, with slight modifications.
+    // The origials can be found here: https://github.com/couchbase/query/blob/master/expression/func_date.go.
+    // This is important, because it means we get the same results as Server N1QL for our date manipulation functions.
+    // I (Callum Birks) have tried rewriting parts of this file to use the date library rather than custom functions,
+    // but it's not really possible while keeping the results identical to Server N1QL. I would advise the future reader
+    // against attempting to rewrite this file.
+
+    // Which quarter this date is in (Q1, Q2, ...)
+    inline int64_t getQuarter(const DateTime& t) { return (t.M + 2) / 3; }
+
+    // The number of leap years up until the given year
+    inline int64_t leapYearsWithin(int64_t year) {
+        if ( year > 0 ) {
+            year--;
+        } else {
+            year++;
+        }
+
+        return (year / 4) - (year / 100) + (year / 400);
+    }
+
+    // The number of leap years between the given years
+    inline int64_t leapYearsBetween(const int64_t start, const int64_t end) {
+        return leapYearsWithin(start) - leapYearsWithin(end);
+    }
+
+    // The Day Of Year for the given time_point. This is the number of days since the start of the year.
+    inline int64_t doy(const system_clock::time_point& t) {
+        const auto daypoint = floor<days>(t);
+        const auto ymd      = year_month_day{daypoint};
+        const auto year     = ymd.year();
+        const auto year_day = daypoint - sys_days{year / January / 0};
+        return year_day.count();
+    }
+
+    inline system_clock::time_point to_time_point(DateTime& dt, bool no_tz = false) {
+        const auto millis = ToMillis(dt, no_tz);
+        return system_clock::time_point(milliseconds(millis));
+    }
+
+    inline bool parseDateArg(sqlite3_value* arg, int64_t* outTime) {
+        const auto str = stringSliceArgument(arg);
+        return str && kInvalidDate != (*outTime = ParseISO8601Date(str));
+    }
+
+    inline bool parseDateArgRaw(sqlite3_value* arg, DateTime* outTime) {
+        if ( sqlite3_value_type(arg) != SQLITE_TEXT ) return false;
+        const auto str = valueAsStringSlice(arg);
```

Comment:
You could replace both lines with a call to `stringSliceArgument`.

Reply:
Ahhh! I would just change it, but I don't want to wait another day for Jenkins failurs 😭